src/main/java/wooteco/subway/domain/route/Router.java

```diff
@@ -0,0 +1,72 @@
+package wooteco.subway.domain.route;
+
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.shortestpath.DijkstraShortestPath;
+import org.jgrapht.graph.DefaultWeightedEdge;
+import org.jgrapht.graph.WeightedMultigraph;
+import org.springframework.stereotype.Component;
+import wooteco.subway.domain.section.Section;
+import wooteco.subway.domain.station.Station;
+
+@Component
+public class Router {
+
+    public Route findShortestRoute(List<Section> sections, Station source, Station target) {
+        List<Station> stations = extractStations(sections);
+
+        validateSourceAndTargetExist(stations, source, target);
+
+        Optional<Route> route = findShortestRoute(stations, sections, source, target);
+
+        return route.orElseThrow(() -> new IllegalStateException("출발지부터 도착지까지 구간이 연결되어 있지 않습니다."));
+    }
+
+    private List<Station> extractStations(List<Section> sections) {
+        return sections.stream()
+                .flatMap(section -> Stream.of(section.getUpStation(), section.getDownStation()))
+                .distinct()
+                .collect(Collectors.toList());
+    }
+
+    private void validateSourceAndTargetExist(List<Station> stations, Station source, Station target) {
+        if (!stations.contains(source) || !stations.contains(target)) {
+            throw new IllegalArgumentException("출발지 또는 도착지에 대한 구간 정보가 없습니다");
+        }
+    }
+
+    private Optional<Route> findShortestRoute(List<Station> stations, List<Section> sections,
+                                              Station source, Station target) {
+        WeightedMultigraph<Station, DefaultWeightedEdge> graph = new WeightedMultigraph<>(DefaultWeightedEdge.class);
+        addStations(stations, graph);
+        addSections(sections, graph);
+
+        DijkstraShortestPath<Station, Integer> dijkstraShortestPath = new DijkstraShortestPath(graph);
```

Comment:
```suggestion
        DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath = new DijkstraShortestPath<>(graph);
```
[Raw Type을 사용하면 안 되는 이유](https://mongsil1025.github.io/book/effective-java/item26/)에 대하여 알아볼까요? 지금 코드는 컴파일 타임에 잡히지 않는 런타임 에러가 발생할 가능성이 있어보여요.

Reply:
범블비 감사합니다!!
라이브러리에 대한 이해가 많이 부족해서 타입 파라미터를 미숙하게 전달했던 것 같습니다.
말씀해주신 것처럼 컴파일 타임에 잡히지 않는 런타임 에러의 발생 가능성은 정말 치명적인 이슈라고 생각합니다!

현재는 다음과 같이 개선해서, 지나온 구간들에 대한 정보도 쉽게 다룰 수 있게 되었습니다!

```java
WeightedMultigraph<Station, SectionEdge> graph = new WeightedMultigraph<>(SectionEdge.class);
addStations(sections.getStations(), graph);
addSections(sections.getSections(), graph);

final GraphPath<Station, SectionEdge> path = new DijkstraShortestPath<>(graph).getPath(source, target);
```

```java
public class SectionEdge extends DefaultWeightedEdge {

    private final Long lineId;
    private final double distance;

    // constructor, getter    
}
```